### PR TITLE
Add Klein-Gordon CCE evolution component, metavariable, and executable 

### PR DESCRIPTION
**Pull request by Sizheng-Ma**
## Proposed changes

This PR adds a metavariable, executable, and evolution component for the Klein-Gordon CCE system. For the moment, the evolution component is almost empty. Actions will be added in the following PRs.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/Systems/Cce/CMakeLists.txt [hunk 1]
```diff
@@ -47,6 +47,7 @@ spectre_target_headers(
   SpecBoundaryData.hpp
   SwshDerivatives.hpp
   System.hpp
+  KleinGordonSystem.hpp
```
Comment (knelli2) [inline src/Evolution/Systems/Cce/CMakeLists.txt:50]: It seems like this was forgotten in a previous PR, so move this to it's own commit

File: src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <type_traits>
+
+#include "Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp"
+#include "Evolution/Systems/Cce/KleinGordonSystem.hpp"
```
Comment (knelli2) [inline src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:9]:
```cpp
#include "Parallel/GlobalCache.hpp"
#include "Parallel/Local.hpp"
#include "Parallel/Phase.hpp"
#include "Time/Actions/SelfStartActions.hpp"
#include "Utilities/TMPL.hpp"
```

File: src/Evolution/Executables/Cce/CMakeLists.txt [hunk 3]
```diff
@@ -16,11 +16,11 @@ set(LIBS_TO_LINK
   Utilities
   )
 
-function(add_cce_executable EXECUTABLE BOUNDARY_COMPONENT)
+function(add_cce_executable EXECUTABLE BOUNDARY_COMPONENT SYSTEM)
   add_spectre_executable(
     ${EXECUTABLE}
     EXCLUDE_FROM_ALL
-    CharacteristicExtract.cpp
+    ${SYSTEM}
```
Comment (knelli2) [inline src/Evolution/Executables/Cce/CMakeLists.txt]: Maybe do
```cpp
${SYSTEM}.cpp
```

so you don't always have to add the `.cpp` in the function calls below.

File: src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Evolution/Executables/Cce/KleinGordonCharacteristicExtract.hpp"
+
+#include <vector>
+
+#include "Evolution/Systems/Cce/Initialize/RegisterInitializeJWithCharm.hpp"
```
Comment (knelli2) [inline src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.cpp:8]:
```cpp
#include "Evolution/Systems/Cce/WorldtubeBufferUpdater.hpp"
#include "Evolution/Systems/Cce/WorldtubeDataManager.hpp"
```

File: src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.hpp [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+#include "Time/Triggers/TimeTriggers.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+template <template <typename> class BoundaryComponent>
+struct EvolutionMetavars : CharacteristicExtractDefaults<false> {
```
Comment (knelli2) [inline src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.hpp:36]: Have you tried compiling this executable yet?

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp (hunk 2 above)
Comment (Sizheng-Ma) [reply src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:9]: Out of curiosity, these header files are already included in `Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp`. I think it should be fine to omit them here, don't you think?

File: src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.hpp (hunk 5 above)
Comment (Sizheng-Ma) [reply src/Evolution/Executables/Cce/KleinGordonCharacteristicExtract.hpp:36]: Yes, it works. But thinking more carefully about it, I feel it's more convenient to include all the actions of `CharacteristicEvolution<Metavariables>` in this commit, since the GR (tensor) part of the two executables should be the same. Actions for the scalar part will be added later. I've just pushed one more commit that does this. For the moment, the executable `KleinGordonCharacteristicEvolution` does the same thing as `CharacteristicEvolution`, except that `Actions::RequestBoundaryData`, `Actions::ReceiveWorldtubeData`, `Actions::RequestNextBoundaryData` are not added, because they are not ready for the `KleinGordonH5WorldtubeBoundary` boundary component. If you are happy with this change, I will squash it with the second commit.

File: src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp (hunk 2 above)
Comment (knelli2) [reply src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:9]: Technically, yes, it's ok to omit them here because they are included in the other `hpp`. However, if somebody in the future decided to remove one of them from `CharacteristicEvolution.hpp`, then this file would fail to compile and somebody would have to spend time trying to figure out why.

So as a rule of thumb, we just always include all files that we need. Even if they may have been included via another file.

File: src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp [hunk 6]
```diff
@@ -51,14 +51,65 @@ struct KleinGordonCharacteristicEvolution
...
+      // note that the initialization will only actually happen on the
+      // iterations immediately following restarts
+      Actions::InitializeFirstHypersurface<
+          evolve_ccm, typename Metavariables::cce_boundary_component>,
+      tmpl::conditional_t<
+          tt::is_a_v<AnalyticWorldtubeBoundary,
+                     typename Metavariables::cce_boundary_component>,
+          Actions::UpdateGauge<false>, Actions::UpdateGauge<evolve_ccm>>,
+      Actions::PrecomputeGlobalCceDependencies,
+      tmpl::conditional_t<evolve_ccm,
```
Comment (knelli2) [inline src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:73]: Do you plan on ever doing a KG CCM system?

**Review by knelli2 — COMMENTED** [review]
Just one question

File: src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp (hunk 6 above)
Comment (Sizheng-Ma) [reply src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:73]: I don't have a plan for the next few months, but I can see two things that could motivate the KG CCM system. One is to simulate ringdown tails. The tail in the tensor sector (i.e., strain and $\Psi_4$) decays faster than the one in the scalar field. Therefore, digging out tails from the scalar field should be much easier than the tensor fields. If it turns out the SpEC CCM has a hard time doing the tensor tail, then I may have to switch to the scalar field first. 

The other one is related to Niko's worldtube projection. I have been talking with him about using this KG CCE system to extract GWs at future null infinity. But the usefulness of the CCM remains unclear.

So I feel it's good to keep those options in case they are needed.

Comment (knelli2) [reply src/Evolution/Systems/Cce/Components/KleinGordonCharacteristicEvolution.hpp:73]: Ok that sounds like  good plan. Let's keep these compile-time switches in here, but they probably won't be used for a while.

You can squash your commits.

**Comment by Sizheng-Ma** [comment]
Done!

**Review by knelli2 — APPROVED** [review]

**Comment by knelli2** [comment]
Test failure unrelated
